### PR TITLE
DM-28429: Convert MultiProFitTask into gen3 PipelineTask

### PR DESCRIPTION
**Pull request by taranu**

### CODE REVIEW

File: fit_multiband.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+import lsst.daf.butler as dafButler
+import lsst.pex.config as pexConfig
+import lsst.pipe.base as pipeBase
+import lsst.pipe.base.connectionTypes as cT
+from typing import Dict, Iterable, List, NamedTuple, Union
+
+__all__ = ["CatalogExposure", "MultibandFitSubTask", "MultibandFitConfig", "MultibandFitTask"]
+
+
+class CatalogExposure(NamedTuple):
```
Comment (TallJimbo) [inline fit_multiband.py]: If you don't need to iterate over these or access them by index, I'd recommend `dataclasses.dataclass` (a decorator) rather than `NamedTuple`.

File: fit_multiband.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+import lsst.afw.image as afwImage
+import lsst.afw.table as afwTable
+import lsst.daf.butler as dafButler
+import lsst.pex.config as pexConfig
+import lsst.pipe.base as pipeBase
+import lsst.pipe.base.connectionTypes as cT
+from typing import Dict, Iterable, List, NamedTuple, Union
+
+__all__ = ["CatalogExposure", "MultibandFitSubTask", "MultibandFitConfig", "MultibandFitTask"]
```
Comment (TallJimbo) [inline fit_multiband.py]: `__all__` should go before your imports (except for `__future__` imports).  You may see a lot of code that puts it here, instead, but only because an old version of flake8 once incorrectly demanded that.

File: fit_multiband.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+
+class CatalogExposure(NamedTuple):
+    band: str
+    catalog: afwTable.SourceCatalog
+    exposure: afwImage.Exposure
+    calib: afwImage.PhotoCalib = None
+    dataId: Union[dafButler.DataCoordinate, None] = None
```
Comment (TallJimbo) [inline fit_multiband.py]: I'd recommend `typing.Optional[T]` instead of `Union[T, None]`, even though they mean the same thing.

File: fit_multiband.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+class MultibandFitConfig(
+    pipeBase.PipelineTaskConfig,
+    pipelineConnections=MultibandFitConnections,
+):
+    fit_multiband = pexConfig.ConfigurableField(
+        target=MultibandFitSubTask,
+        doc="Task to fit sources using multiple bands",
+    )
+
+    def get_band_sets(self):
```
Comment (TallJimbo) [inline fit_multiband.py]: Needs docstring.

File: fit_multiband.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+        dict_fit = self.toDict()['fit_multiband']
+        bands_to_fit = dict_fit.get('bands_fit')
+        if bands_to_fit is None:
+            raise RuntimeError(f'{__class__}.fit_multiband must have bands_fit attribute '
```
Comment (TallJimbo) [inline fit_multiband.py]: I would not have thought `__class__` alone would work here.  Am I just learning something new, or should this be `self.__class__`?

File: fit_multiband.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    pipelineConnections=MultibandFitConnections,
+):
+    fit_multiband = pexConfig.ConfigurableField(
+        target=MultibandFitSubTask,
+        doc="Task to fit sources using multiple bands",
+    )
+
+    def get_band_sets(self):
+        dict_fit = self.toDict()['fit_multiband']
+        bands_to_fit = dict_fit.get('bands_fit')
```
Comment (TallJimbo) [inline fit_multiband.py]: I think I'd recommend
```
try:
    self.fit_multiband.bands_fit
except AttributeError:
    raise RuntimeError(...) from None
```
(etc)

I'm all for using `dict.get` when the missing case is not an error, but it is an error here, and that's exactly what exceptions are for (also, calling `toDict` is going to convert a lot of stuff to dict form that you don't need).

File: fit_multiband.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+                elif bands_set != bands_data:
+                    raise ValueError(f'Datatype={type_d} bands_set={bands_set}) != previous={bands_data})')
+                if has_extra_bands:
+                    for dref in ref_d:
+                        if dref.dataId['band'] not in bands_needed:
+                            ref_d.remove(dref)
+        return datasetRefMap
+
+
+class MultibandFitSubTask(pipeBase.Task):
```
Comment (TallJimbo) [inline fit_multiband.py]: You might be able to make this multiply-inherit from `abc.ABC`, too, which would let you make use `@abstractmethod` on `run`.  (I think I've seen a `Task` do that, but if it doesn't work because of some nonsense in the base class, don't worry about it).

In any case, :+1: to having a base class to define the interface from me, even if it's just a place to hang docs.

File: fit_multiband.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+        schema : `lsst.afw.table.Schema`
+            The input schema for the reference source catalog, used to initialize the output schema.
+        kwargs
+            Additional arguments to be passed to the subclass constructor(s).
+
+        Notes
+        -----
+        This method must initialize an attribute named schema with the schema of the output catalog.
+        """
+        raise RuntimeError("Not implemented")
```
Comment (TallJimbo) [inline fit_multiband.py]: This needs to call `super().__init__(**kwargs)` or it won't get its logger, metadata, etc. set up properly.  And subclass `__init__` methods then need to call `super().__init__(schema, **kwargs)` to do the same, so it shouldn't raise.

File: fit_multiband.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+        """ Fit sources from a reference catalog using data from multiple exposures in the same region (patch).
+
+        Parameters
+        ----------
+        catexps : `typing.List [CatalogExposure]`
+            A list of catalog-exposure pairs in a given band. Subclasses may require there to only be one catexp per
+            band, and/or for the catalogs to contain HeavyFootprints with deblended images.
+        cat_ref : `lsst.afw.table.SourceCatalog`
+            A reference source catalog to fit. Subclasses may be configured fit only a subset of these sources.
+        kwargs
```
Comment (TallJimbo) [inline fit_multiband.py]:
```suggestion
        **kwargs
```

Comment (TallJimbo) [reply fit_multiband.py]: If these kwargs are different for each subtask implementation, it's not really a well-defined interface, because there's no way for the parent task to know what the subtask expects.  I hope they're at least never required, and if you can find a way to get rid of them entirely, that might be better.

Like with the `__init__` kwargs, if some derived class wants to add its own additional keyword arguments that _can_ be used when someone is using that class directly (i.e. not through the parent task), they can do that even if the base class method does not define `**kwargs`, as long as the derived-class method still accepts all of the arguments the base class method does.

File: fit_multiband.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+            Additional keyword arguments.
+
+        Returns
+        -------
+        retStruct : `lsst.pipe.base.Struct`
+            A struct with a cat_output attribute containing the output measurement catalog.
+        """
+        raise RuntimeError("Not implemented")
```
Comment (TallJimbo) [inline fit_multiband.py]: Better to raise `NotImplementedError` here; that's exactly what it's for.

File: fit_multiband.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+        ----------
+        schema : `lsst.afw.table.Schema`
+            The input schema for the reference source catalog, used to initialize the output schema.
+        kwargs
+            Additional arguments to be passed to the subclass constructor(s).
+
+        Notes
+        -----
+        This method must initialize an attribute named schema with the schema of the output catalog.
+        """
```
Comment (TallJimbo) [inline fit_multiband.py]: `__init__` doctrings should actually [go in the class docstring](https://developer.lsst.io/python/numpydoc.html#py-docstring-class-structure).  You should also add a note there that all subclasses must have the same `__init__` signature, because that isn't usually something base classes demand (especially because it's the only reason this abstract `Task` even defines `__init__`).

File: fit_multiband.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+        bands_needed = bands_fit.union(bands_read)
+
+        bands_data = None
+        has_extra_bands = False
+
+        for type_d, ref_d in datasetRefMap.items():
+            # Datasets without bands in their dimensions should be fine
+            if 'band' in type_d.dimensions:
+                bands = [dref.dataId['band'] for dref in ref_d]
+                bands_set = set(bands)
```
Comment (TallJimbo) [inline fit_multiband.py]: You can write these two lines as a `set` comprehension:
```
bands = {dref.dataId['band'] for dref in ref_d}
```

File: fit_multiband.py [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+                if bands_data is None:
+                    bands_data = bands_set
+                    if bands_data != bands_fit:
+                        if not bands_needed.issubset(bands_data):
+                            raise ValueError(f'bands_needed={bands_needed}) !subsetof bands_set={bands_set}'
+                                             f' from refs={ref_d})')
+                        has_extra_bands = True
+                elif bands_set != bands_data:
+                    raise ValueError(f'Datatype={type_d} bands_set={bands_set}) != previous={bands_data})')
```
Comment (TallJimbo) [inline fit_multiband.py]: Given that it will be easy for users to run into these exceptions, they're probably places where you want really verbose messages that tell them exactly how to fix them, i.e. which configuration options are bad, and how.

File: fit_multiband.py [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+    ConfigClass = pexConfig.Config
+
+    def __init__(self, schema: afwTable.Schema, **kwargs):
+        """ Initialize the task, defining and setting the output catalog schema.
+
+        Parameters
+        ----------
+        schema : `lsst.afw.table.Schema`
+            The input schema for the reference source catalog, used to initialize the output schema.
+        kwargs
```
Comment (TallJimbo) [inline fit_multiband.py]:
```suggestion
        **kwargs
```

File: fit_multiband.py [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+            Additional arguments to be passed to the subclass constructor(s).
```
Comment (TallJimbo) [inline fit_multiband.py]: I think you mean base `Task` constructor; see other comment on the `__init__` body, below.

If you _also_ meant this to be a way for subclasses to customize their construction, well, I think they can't be allowed to, or the parent `Task` that uses them as a subtask won't be able to construct them.  And if those extra per-subclass kwargs are optional, it's okay to add them to the subclass `__init__` without mentioning them here, as long as they're strictly additions in the subclass, not subtractions or modifications.

File: fit_multiband.py [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+        super().__init__(initInputs=initInputs, **kwargs)
+        self.makeSubtask("fit_multiband", schema=initInputs["cat_ref_schema"].schema)
+        self.cat_output_schema = afwTable.SourceCatalog(self.fit_multiband.schema)
+
+    def runQuantum(self, butlerQC, inputRefs, outputRefs):
+        inputs = butlerQC.get(inputRefs)
+        input_objs = [inputRefs.cats_meas, inputRefs.coadds]
+        dataIds_cats, dataIds = [[dRef.dataId for dRef in obj] for obj in input_objs]
+        if not all((x == y for x, y in zip(dataIds, dataIds_cats))):
+            raise RuntimeError(f'Mismatched cats_meas, coadds dataIds {dataIds_cats} {dataIds}')
```
Comment (TallJimbo) [inline fit_multiband.py]: I don't think consistent ordering is guaranteed by the caller here.  Better to use dicts and reorder defensively.

If you want to check that the same data IDs are present for both dataset types, I think you can do that in `adjustQuantum`, maybe?

File: fit_multiband.py [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+        if bandset_data != bandset_needed:
+            raise RuntimeError(f'set(band={bands_data})={bandset_data} != bandset_needed={bandset_needed}'
+                               f'=(bands_fit={bandset_fit}).union(bands_read={bandset_read}))')
+        cat_ref, cats_meas, coadds = (inputs[x] for x in ('cat_ref', 'cats_meas', 'coadds'))
+        n_data = [len(x) for x in (bands_data, cats_meas, coadds)]
+        if not all(x == n_data[0] for x in n_data[1:]):
+            raise RuntimeError(f'bands, cats_meas, coadds, dataIds lens={n_data} not matched')
+        catexps = [
+            CatalogExposure(band=band, catalog=cat, exposure=exp, dataId=dataId, metadata={},
+                            calib=exp.getPhotoCalib() if hasattr(exp, 'getPhotoCalib') else None)
```
Comment (TallJimbo) [inline fit_multiband.py]: If `exp` is an `afw.image.Exposure`, that `hasattr` test will always succeed, even if the method _returns_ `None`.  On that note, do you need a separate `PhotoCalib` field in `CatalogExposure` anyway, if you just get it from the exposure?

File: fit_multiband.py [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+        bandset_fit, bandset_read = self.config.get_band_sets()
+        bandset_needed = bandset_fit.union(bandset_read)
+        bandset_data = set(bands_data)
+        if bandset_data != bandset_needed:
+            raise RuntimeError(f'set(band={bands_data})={bandset_data} != bandset_needed={bandset_needed}'
+                               f'=(bands_fit={bandset_fit}).union(bands_read={bandset_read}))')
+        cat_ref, cats_meas, coadds = (inputs[x] for x in ('cat_ref', 'cats_meas', 'coadds'))
+        n_data = [len(x) for x in (bands_data, cats_meas, coadds)]
+        if not all(x == n_data[0] for x in n_data[1:]):
+            raise RuntimeError(f'bands, cats_meas, coadds, dataIds lens={n_data} not matched')
```
Comment (TallJimbo) [inline fit_multiband.py]: Some comments around all of the logic in this method would be welcome; I could not keep track of what the different lists/sets of bands really meant or how you were relating them.

And if any of these checks are invariants that are guaranteed by `adjustQuantum`, feel free to just assert on them (or even skip them) here.  If they are not guaranteed by `adjustQuantum`, please add some context to the error messages.

File: fit_multiband.py [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+        Parameters
+        ----------
+        schema : `lsst.afw.table.Schema`
+            The input schema for the reference source catalog, used to initialize the output schema.
+        kwargs
+            Additional arguments to be passed to the subclass constructor(s).
+
+        Notes
+        -----
+        This method must initialize an attribute named schema with the schema of the output catalog.
```
Comment (TallJimbo) [inline fit_multiband.py]: If you can make this an ABC, you can do
```
@property
@abstractmethod
def schema(self) -> afwTable.Schema:
    raise NotImplementedError()
```
if you don't mind forcing subclasses to use a property - it's a bit more pain for subclasses, but you get some nice declarative clarity.  Your call.

File: fit_multiband.py [hunk 20]
```diff
@@ -0,0 +1,255 @@
...
+        cat_ref : `lsst.afw.table.SourceCatalog`
+            A reference source catalog to fit. Subtasks may be configured fit only a subset of these sources.
+
+        Returns
+        -------
+        retStruct : `lsst.pipe.base.Struct`
+            A struct with a cat_output attribute containing the output measurement catalog.
+        """
+        cat_output = self.fit_multiband.run(catexps, cat_ref).output
+        # It would be best to validate this earlier but I'm not sure how
```
Comment (TallJimbo) [inline fit_multiband.py]: Not much you can do; this is something subclasses are expected to just be responsible for.

But you might want to move this check to `runQuantum`, and do it _after_ the `put`, so if there is a problem, the user ends up with files they can go look at to see how they're wrong.  It'd be nice to also have it here in `run`, but in practice I think it'll be a rare enough problem that making it easy to debug when it happens is more important.

File: fit_multiprofit.py [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+    dataId_band = dataId.copy()
+    dataId_band['band'] = band
+    dataIds_band.append(dataId_band)
+
+ids = dict(cat_ref=dataId, coadds=dataIds_band, cats_meas=dataIds_band, cat_output=dataId)
+quantum = makeQuantum(task, butler, dataId, ids)
+
+for mockRun in (True, False):
+    print(f'Running test quantum with mockRun={mockRun}')
+    runTestQuantum(task, butler, quantum, mockRun=mockRun)
```
Comment (TallJimbo) [inline fit_multiprofit.py:78]: Please move scripts like this to your main Python package so they're importable, move logic to a `main()` function or similar, and either just run them with `python -m` or add a minimal launcher script to `bin`.

I see you've got a lot of others in the package root, which are probably reasonably out of scope on this ticket, but best to get this one and future ones into standards conformance.

File: fit_multiprofit_config.py [hunk 22]
```diff
@@ -0,0 +1,12 @@
...
+config.connections.name_output_cat = "multiprofit"
+config.fit_multiband.retarget(MultiProFitTask)
+config.fit_multiband.bands_read = ["r"]
+config.fit_multiband.fitCModelExp = False
+config.fit_multiband.fitSersicFromCModel = True
+config.fit_multiband.computeMeasModelfitLikelihood = True
+config.fit_multiband.usePriorShapeDefault = True
+config.fit_multiband.priorCentroidSigma = 0.2
+config.fit_multiband.priorMagBand = "r"
+config.fit_multiband.psfHwhmShrink = 0.1
```
Comment (TallJimbo) [inline fit_multiprofit_config.py:11]: Should these defaults just live in `MultiProFitTask`'s config class?

File: multiprofit_task.py [hunk 23]
```diff
@@ -56,9 +87,12 @@ class MultiProFitConfig(pexConfig.Config):
...
                                       "contiguous positive pixels in a naively de-noised image")
-    filenameOut = pexConfig.Field(dtype=str, default="", doc="Filename for output of FITS table")
-    filenameOutDeblend = pexConfig.Field(dtype=str, default="", doc="Filename for output of FITS table with"
-                                                                    " deblended fits")
+    field_localbg = pexConfig.Field(
+        dtype=str, default='base_LocalBackground', doc="Field name (prefix to _instFlux) to read the local background "
+                                                       "level from if usePriorBackgroundLocalEstimate is True")
+    filenameOut = pexConfig.Field(dtype=str, optional=True, doc="Filename for output of FITS table")
+    filenameOutDeblend = pexConfig.Field(dtype=str, optional=True, doc="Filename for output of FITS table with"
+                                                                      " deblended fits")
```
Comment (TallJimbo) [inline multiprofit_task.py:91]: The code that uses these options should never be run inside a `PipelineTask`, because all I/O needs to go through butler.  If it's hard to factor them out of the subtask that the `PipelineTask` will use, perhaps the base subtask interface should define a method like
```
@classmethod
def validateConfigsForPipelineTask(cls, config: ConfigClass) -> None:
    """Raise if the given config includes settings that are illegal when used in a PipelineTask.
    """.
    pass  # assume most subclasses don't have any such options
```
This task could then implement that to check that `config.filenameOut is None` (etc), and the parent `PipelineTask` could call it.

File: multiprofit_task.py [hunk 24]
```diff
@@ -130,11 +166,19 @@ class MultiProFitConfig(pexConfig.Config):
...
     outputLogLikelihood = pexConfig.Field(dtype=bool, default=True, doc="Whether to save the log likelihood "
                                                                         "of each model's best fit")
-    outputRuntime = pexConfig.Field(dtype=bool, default=True, doc="Whether to save the runtime of each "
-                                                                  "model")
+    outputRuntime = pexConfig.Field(dtype=bool, default=True, doc="Whether to save the runtime of each model")
+    pathCosmosGalsim = pexConfig.Field(
+        dtype=str, optional=True,
+        doc="Path containing COSMOST HST F814W-based real_galaxy_catalog_25.2.fits and "
+            "real_galaxy_PSF_images_25.2_n[1-88].fits (https://zenodo.org/record/3242143)"
+    )
```
Comment (TallJimbo) [inline multiprofit_task.py:170]: Another option that looks like it involves behavior that isn't legal in a `PipelineTask`.

**Review by TallJimbo — COMMENTED** [review]
Lots of minor comments on the `PipelineTask`; structurally it all looks good, but there are:
- some bits that are hard to follow and could use a lot of comments (e.g. I'm not sure I followed `adjustQuantum` well enough to answer your question about that from Jira, but I'll give it a go back on Jira anyway, to keep that conversation together);
- a lot of error messages that look very useful to you, but pretty opaque to anyone else;
- a lot of scripts that should be modules.

Not all of those should be fixed on this ticket, as it's mostly preexisting, but taking some steps in that direction in the new code will help us get at least some of this into lsst_distrib and regular code review someday.

File: fit_multiband.py (hunk 1 above)
Comment (taranu) [reply fit_multiband.py]: I wanted the non-metadata fields to be immutable, but now I see I can do that with a frozen dataclass and set the default metadata to a mutable empty dict as well. Thanks.

File: fit_multiband.py (hunk 5 above)
Comment (taranu) [reply fit_multiband.py]: https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy Further down under "Class instances" where there are sadly no anchors, it says: "Special attributes: `__dict__` is the attribute dictionary; `__class__` is the instance’s class.", though I doubt I really knew that so much as my IDE didn't flag it.

Comment (TallJimbo) [reply fit_multiband.py]: Right; those are instance _attributes_, which means you need to do something like `self.__class__` to access them.

Comment (taranu) [reply fit_multiband.py]: Apparently you don't and `__class__` just works as per the docs for reasons I don't entirely understand but am willing to accept.

File: fit_multiprofit.py (hunk 21 above)
Comment (taranu) [reply fit_multiprofit.py:78]: I don't plan to get this script and basically anything in `modelling_research` (besides the MultiProFitTask) into a stack package or into compliance anytime soon, if at all. This script was essentially my trying to work out what an activator does. I suppose in the future I'll try to make runnable scripts importable with a `main()`.

File: fit_multiprofit_config.py (hunk 22 above)
Comment (taranu) [reply fit_multiprofit_config.py:11]: They already have defaults; these are my currently-favoured overrides that I'm not confident enough to make into defaults.

I'm not sure if it's generally a good idea to have a default prior band, although in this case I haven't actually implemented a prior based on any band besides `r` (yet).

File: multiprofit_task.py (hunk 23 above)
Comment (taranu) [reply multiprofit_task.py:91]: I intend to refactor the deblending-from-previous-single-source-fitting-catalogs into a separate `PipelineTask` in another ticket; it's too substantial of a change to include here.

File: multiprofit_task.py (hunk 24 above)
Comment (taranu) [reply multiprofit_task.py:170]: To be honest, I haven't used this option in years and I'm not sure it even works anymore, so I'll probably just remove it. The other `fitHstCosmos`-related options also do I/O through some cutout code, which you wouldn't know without digging further into the relevant functions.
This isn't urgent but are there legal ways of using external catalogs/images, or of ingesting say, coadds & PSFs of arbitrary provenance (like COSMOS-HST?).

Comment (TallJimbo) [reply multiprofit_task.py:170]: If you can ingest them into the butler, then it becomes legal.  Whether that's hard depends largely on whether you can easily cook up a data ID for them that only uses our existing data ID key-value pairs ("no data ID because there can only be one" may be a valid option).
